vortex-file/src/read/buffered.rs

```diff
@@ -0,0 +1,232 @@
+            .cache
+            .write()
+            .unwrap_or_else(|poison| vortex_panic!("Failed to write to message cache: {poison}"));
+        for Message(message_id, buf) in messages {
+            write_cache_guard.set(message_id, buf);
+        }
+    }
+
+    fn gather_read_messages(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> VortexResult<(Vec<MessageLocator>, bool)> {
+        let mut to_read = Vec::with_capacity(NUM_TO_COALESCE);
+        let mut read_more_count = 0;
+
+        // Poll all queued pending masks to see if we can make progress
+        for queued_res in self.queued.iter_mut() {
+            match queued_res {
+                RowMaskState::Pending(pending_mask) => {
+                    if let Some(pending_read) = self.row_mask_reader.read_mask(pending_mask)? {
+                        match pending_read {
+                            MessageRead::ReadMore(m) => {
+                                to_read.extend(m);
+                                read_more_count += 1;
+                            }
+                            MessageRead::Value(v) => {
+                                *queued_res = RowMaskState::Ready(v);
+                            }
+                        }
+                    } else {
+                        *queued_res = RowMaskState::Empty;
+                    }
+                }
+                RowMaskState::Ready(_) => {}
+                RowMaskState::Empty => {}
+            }
+        }
+
+        let mut exhausted = false;
+        while read_more_count < NUM_TO_COALESCE {
+            match self.values.poll_next_unpin(cx) {
+                Poll::Ready(Some(Ok(next_mask))) => {
+                    if let Some(read_result) = self.row_mask_reader.read_mask(&next_mask)? {
+                        match read_result {
+                            MessageRead::ReadMore(m) => {
+                                self.queued.push_back(RowMaskState::Pending(next_mask));
+                                to_read.extend(m);
+                                read_more_count += 1;
+                            }
+                            MessageRead::Value(v) => {
+                                self.queued.push_back(RowMaskState::Ready(v));
+                            }
+                        }
+                    }
+                }
+                Poll::Ready(Some(Err(e))) => {
+                    return Err(e);
+                }
+                Poll::Ready(None) => {
+                    exhausted = true;
+                    break;
+                }
+                Poll::Pending => {
+                    break;
+                }
+            }
+        }
+        Ok((to_read, exhausted))
+    }
+
+    fn dispatch_messages(
+        &self,
+        messages: Vec<MessageLocator>,
+        waker: Waker,
+    ) -> BoxFuture<'static, io::Result<Vec<Message>>> {
+        let reader = self.io_read.clone();
+        self.dispatcher
```

Comment:
I don't think you need this outer `dispatcher.dispatch`, since `read_byte_ranges` is also going through the Dispatcher.

Reply:
I had a deadlock without it. The await is never resumed as we propagate the pending to the stream but if stream returns pending it needs to wake itself up so I need something to poll the read and then call wake